baseapp/baseapp.go

```diff
@@ -771,25 +811,50 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
 // and DeliverTx. An error is returned if any single message fails or if a
 // Handler does not exist for a given message route. Otherwise, a reference to a
 // Result is returned. The caller must not commit state if an error is returned.
-func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode) (*sdk.Result, error) {
+func (app *BaseApp) runMsgs(ctx sdk.Context, msgs []sdk.Msg, mode runTxMode, isNonAtomic bool) (*sdk.Result, error) {
 	events := sdk.EmptyEvents()
 	var msgResponses []*codectypes.Any
 
+	txSuccess := false
+	msgCtx := ctx // This context may be replaced on each message for non-atomic execution.
+
 	// NOTE: GasWanted is determined by the AnteHandler and GasUsed by the GasMeter.
 	for i, msg := range msgs {
 		if mode != runTxModeDeliver && mode != runTxModeSimulate {
 			break
 		}
 
+		// When dealing with non-atomic execution, we want each message to have its own context
+		var writeFn func()
+		if isNonAtomic {
+			msgCtx, writeFn = ctx.CacheContext()
+		}
+
 		handler := app.msgServiceRouter.Handler(msg)
 		if handler == nil {
 			return nil, errorsmod.Wrapf(sdkerrors.ErrUnknownRequest, "can't route message %+v", msg)
 		}
 
 		// ADR 031 request type routing
-		msgResult, err := handler(ctx, msg)
+		msgResult, err := handler(msgCtx, msg)
 		if err != nil {
-			return nil, errorsmod.Wrapf(err, "failed to execute message; message index: %d", i)
+			wrappedErr := errorsmod.Wrapf(err, "failed to execute message; message index: %d", i)
```

Comment:
Unfortunately, this is unsafe. The block header includes the following information as determinstic:
* Code 
* Data (marshalled msg responses)
* GasWanted
* GasUsed

This combination creates a potential halting scenario. This code will include the full error string from the msg exeuction into the msg responses, thus causing the error string to know be included in the block. If the error string is undeterminstic (happens with json marshal errors) or it is changed (a backport changes error string formatting), a chain would halt. We have this same issue with writing error strings into acknowledgements for [IBC](https://github.com/cosmos/ibc-go/blob/main/modules/core/04-channel/types/acknowledgement.go#L35), which is common occurrence with interchain accounts 

To increase the readability of these error messages, I'd like to ask cometbft to include the codespace in the block header as well (I will open an issue), but I also think there should potentially be a more radical change to include the base error message as well. The original design was to use the Code as a condensed error string with the idea that that clients would lookup the code to get a more descriptive error string, but this is infeasible with IBC as a client would need to look up a different chain's error code

Reply:
This makes sense. A simple solution would be to extract error code via `errorsmod.ABCIInfo`

Reply:
the logic which is unsafe here seems to be the same as it was before, id opt to not return any sort of error to comet which gets included in consensus, (wiping the included data on errors that gets included in the header). I think this is fundamentally unsafe and its surprising no one has halted because of this.

Reply:
The difference between before and after is that now the error is being included in the msgResponses which will eventually be marshalled and placed into the `Data` field (which is a determinsitic field). 

Existing flow:
- [error returned on failed msg]()
- [which is passed back from runTx](https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/baseapp.go#L740)
- [which is put into a deliver tx response](https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/abci.go#L395)
- [and the non-deterministic error is removed](https://github.com/cosmos/cosmos-sdk/blob/main/types/errors/abci.go#L11)

Thus the non-determinstic parts of an error returned on a failed msg execution are removed for the determinstic fields of the deliver tx response. In the new flow, the full error is put into the msg response below on line 843, which eventually ends up in the Data field which could lead to an app hash mismatch

One tricky part is that you likely want the errors to end up in the deliver tx response as [well](https://github.com/cosmos/cosmos-sdk/blob/main/errors/abci.go#L37) since the log is non-determinstic at the moment and can provide useful information

Reply:
I think having the error that gets included in the error responses be just the ABCIInfo end [emit the context as events](https://github.com/cosmos/cosmos-sdk/issues/16174) or logs would solve this.

Also, hot take: consensus failures on error string differences are should be irrelevant as long as all validators return an error, and there should be a way to fix this post-hoc at a performance penalty.

Reply:
Made the change so that errors are converted to ABCI errors in [5a0ffba](https://github.com/cosmos/cosmos-sdk/pull/15038/commits/5a0ffbae573815cdac3ad2a36386ac9cf3e9aff5)